Mathlib/Topology/Order/ScottTopology.lean

```diff
@@ -0,0 +1,400 @@
+import Mathlib.Topology.ContinuousFunction.Basic
+
+/-!
+# Scott topology
+This file introduces the Scott topology on a preorder.
+## Main definitions
+- `preserve_lub_on_directed` - a function between preorders which preserves least upper bounds.
+- `with_scott_topology.topological_space` - the Scott topology is defined as the join of the
+  topology of upper sets and the topological space where a set `u` is open if, when the least upper
+  bound of a directed set `d` lies in `u` then there is a tail of `d` which is a subset of `u`..
+## Main statements
+- `with_scott_topology.is_open_is_upper` - Scott open sets are upper
+- `with_scott_topology.is_closed_is_lower` - Scott closed sets are lower
+- `with_scott_topology.continuous_monotone` - Scott continuous functions are monotone.
+- `preserve_lub_on_directed_iff_scott_continuity` - a function preserves least upper bounds of
+  directed sets if and only if it is Scott continuous
+- `with_scott_topology.t0_space` - the Scott topology on a partial order is T₀
+## Implementation notes
+A type synonym `with_scott_topology` is introduced and for a preorder `α`, `with_scott_topology α`
+is made an instance of `topological_space` by the topology generated by the complements of the
+closed intervals to infinity.
+A class `Scott` is defined in `topology.omega_complete_partial_order` and made an instance of a
+topological space by defining the open sets to be those which have characteristic functions which
+are monotone and preserve limits of countable chains. Whilst this definition of the Scott topology
+coincides with the one given here in some special cases, in general they are not the same
+[Domain Theory, 2.2.4][abramsky_gabbay_maibaum_1994].
+## References
+* [Gierz et al, *A Compendium of Continuous Lattices*][GierzEtAl1980]
+* [Abramsky and Jung, *Domain Theory*][abramsky_gabbay_maibaum_1994]
+## Tags
+Scott topology, preorder
+-/
+
+variable (α β : Type _)
+
+open Set
+
+section preorder
+
+variable {α} {β}
+
+variable [Preorder α] [Preorder β]
+
+lemma is_upper_set_iff_forall_le  {s : Set α} : IsUpperSet s ↔ ∀ ⦃a b : α⦄, a ≤ b →
+  a ∈ s → b ∈ s := Iff.rfl
+
+/--
+The set of upper sets forms a topology
+-/
+def upper_set_topology : TopologicalSpace α :=
+{ IsOpen := IsUpperSet,
+  isOpen_univ := isUpperSet_univ,
+  isOpen_inter := fun _ _ => IsUpperSet.inter,
+  isOpen_unionₛ := fun _ h => isUpperSet_unionₛ h, }
+
+lemma pair_is_chain (a b : α) (hab: a ≤ b) : IsChain (· ≤ ·) ({a, b} : Set α) := by
+  apply IsChain.insert (Set.Subsingleton.isChain subsingleton_singleton)
+  intros c h₁ h₂
+  rw [mem_singleton_iff] at h₁
+  rw [h₁]
+  exact Or.inl hab
+
+lemma directed_on_pair (a b : α) (hab: a ≤ b) : DirectedOn (· ≤ ·) ({a, b} : Set α) :=
+  (pair_is_chain _ _ hab).directedOn
+
+/--
+A function which preserves lub on directed sets
+-/
+def preserve_lub_on_directed (f : α → β) := ∀ (d : Set α) (a : α), d.Nonempty → DirectedOn (· ≤ ·)
+  d → IsLUB d a → IsLUB (f '' d) (f a)
```

Comment:
This belongs in an earlier file.

Reply:
I think that's probably true of everything in the `preorder` section.

Reply:
Also, it should probably be `def preserveLUB_onDirected` or something like that? (Is the new naming convention documented somewhere?)

Reply:
I've split this out into a separate PR (for mathlib since it touches a ported file) https://github.com/leanprover-community/mathlib/pull/18517

Reply:
Now merged, thanks.